ARROW-6074: [FlightRPC][Java] Middleware

lihalite opened the pull request:
(This is on top of #5013 since they would conflict otherwise.)

This is a draft implementation of middleware for Flight in Java, both client-side and server-side, to accompany the corresponding [mailing list discussion](https://lists.apache.org/thread.html/82a7c026ad18dbe9fdbcffa3560979aff6fd86dd56a49f40d9cfb46e@%3Cdev.arrow.apache.org%3E).

In the integration tests, there's an example of a simplified OpenTracing-like integration. This also sends various uncaught exceptions in the Flight implementation to the middleware for handling.

Some unresolved details:
- This allows for both binary and text headers, but this is a bit of a gRPC detail. We may instead to just present one or the other (and fold them into each other, e.g. binary headers get presented as base64-encoded headers).
- At this point, we may want to just expose headers to the RPC method bodies as well (through `CallContext`).
- This does its best to expose an HTTP/1-like request flow where headers always precede bodies, and there is always only one set of headers, even though this is not true of gRPC. (gRPC may skip initial headers if the response is an error, and always sends a set of trailers.)
- The final (or followup PR) should add a complete OpenTracing example somewhere (e.g. in a `contrib/` directory) to ensure that integrating with such APIs works well.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/MetadataAdapter.java:
```diff
@@ -0,0 +1,73 @@
+import io.grpc.Metadata.Key;
+
+/**
+ * A mutable adapter between the gRPC Metadata object and the Flight headers interface.
```
can you document the use-case?

emkornfield commented:
I took a quick look through this, and it seems reasonable, need to do another closer look for details. Two questions:
1.   is there a reason that you accept multiple middleware implementations/factories?  it seems like it might make some of the core code simpler if you only accepted one and provide an implementation as of a composite middleware object.
2.  It would be good to clarify expected thread-safety guarantees for the middleware (sorry if I missed them).

Also note, it looks like a rebase is required.

lihalite commented:
Thanks @emkornfield for taking a look, I've rebased. I added some more comments about thread-safety of server middleware (FlightServerMiddleware.java) and documented the use of MetadataAdapter.

I chose to allow the user to give a list of middleware in order to have a convenient way to access middleware state from within a RPC method implementation. (I chose this approach as I didn't want to tie us to gRPC's context mechanism.) It could be done with a composite middleware object as you suggest, but that would be another interface on top.

emkornfield commented:
>I chose to allow the user to give a list of middleware in order to have a convenient way to access middleware state from within a RPC method implementation.

What are the use cases you had in mind for this?

lihalite commented:
> 
> 
> > I chose to allow the user to give a list of middleware in order to have a convenient way to access middleware state from within a RPC method implementation.
> 
> What are the use cases you had in mind for this?

In this case, passing an OpenTracing span (or other unique request identifier) from the server middleware to client middleware. Another use case might be implementing an auth mechanism without using Flight's mechanism, and passing the identity of the peer to the method body. (Flight doesn't allow, for instance, making auth optional on certain methods.) I'm not attached to this design, though, so long as there's some way to do it. An alternative might be to let middleware update a Map stored in the `CallContext` (perhaps using the typed key pattern for better type safety), so that we're not tied directly to the middleware instances.

gRPC uses a context object, which has to be manually passed around throughout the request (this is mostly invisible to you unless you are a middleware author or you spawn threads inside your RPC method implementation). Data gets passed around by means of static final key objects which get/set a corresponding object inside the context. I don't want to expose this in Flight and don't like the odd global-looking state (i.e. I'd rather make Flight put this in `CallContext` somehow).

lihalite commented:
Rebased, Travis build: https://travis-ci.com/lihalite/arrow/builds/125596743

I would consider this ready to review, with how to expose headers as a question that could be resolved. If adding a `contrib/` directory makes sense, I can add an OpenTracing/OpenTelemetry example as well.

praveenbingo commented:
@emkornfield i am not too familiar with this part of arrow (also we are not yet integrated with flight)

will take a stab at reviewing this though..

lihalite commented:
Thanks @rymurr! Yes, Python/C++ client/servers do not support middleware yet. If a Java middleware adds headers, Python/C++ will ignore them.

lihalite commented:
Just following up here. I've rebased/updated the branch, are there any further comments on this implementation?

emkornfield commented:
@lihalite I should have some time later in the week to do another review if you think it is warranted.

lihalite commented:
@emkornfield the code is the same as when you looked, did I address your questions above sufficiently?

wesm commented:
What is the status of this PR? It seems like we would want to merge this, review and merge the C++/Python PR, then work on reconciling the details and adding integration tests

lihalite commented:
I believe it's ready for review, and so is the C++/Python one. I can look at integration at parallel, I think I commented on JIRA asking whether it would be worth refactoring the current script before we add on a lot of new things to it.

emkornfield commented:
I'll take one more pass through this today.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/CallHeaders.java:
```diff
@@ -0,0 +1,43 @@
+  /** Get the value of a header with a binary encoding. */
+  byte[] getBinary(String key);
+
+  /** Get the value of a header with a text encoding. */
```
how does one know whether a particular key is valid text?

lihalite replied:
Given that the C++/Python implementation doesn't expose this distinction, I'll elide the binary one here. (That should also make it more portable to other protocols.)

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/CallInfo.java:
```diff
@@ -0,0 +1,33 @@
+ * A description of a Flight call for middleware to inspect.
+ */
+public class CallInfo {
+  final String method;
```
private?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/CallInfo.java:
```diff
@@ -0,0 +1,33 @@
+public class CallInfo {
+  final String method;
+
+  public CallInfo(String method) {
```
should method be called methodName?  Is there any contract required on the name?  e.g. Service.MethodName?

lihalite replied:
I've changed this to use an enum instead (to avoid leaking the gRPC method name).

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -77,11 +81,21 @@
-    asyncStub = FlightServiceGrpc.newStub(channel).withInterceptors(authInterceptor);
+    final ClientInterceptor[] interceptors;
+
+    if (middleware.isEmpty()) {
```
is there a reason to specialize here (does having a do nothing ClientInterceptorAdaptor add appreciable overhead?)

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -77,11 +81,21 @@
+
+    blockingStub = FlightServiceGrpc.newBlockingStub(channel).withInterceptors(interceptors);
+    asyncStub = FlightServiceGrpc.newStub(channel).withInterceptors(interceptors);
+    // TODO: doGet/doPut need to run all interceptors
```
can you make a JIRA describing this issue more clearly?

lihalite replied:
I've fixed it (DoGet/DoPut used to manually run the gRPC interceptor chain, I've refactored that).

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClientMiddleware.java:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * A callback used before request headers are sent. The headers may be manipulated.
+   */
+  void sendingHeaders(CallHeaders outgoingHeaders);
```
if this is a callback, a better name might be:  "onSend"?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightClientMiddleware.java:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * A callback called after response headers are received. The headers may be manipulated.
+   */
+  void headersReceived(CallHeaders incomingHeaders);
```
same comment as above on naming, using the "on" prefix I think makes semantics clearer.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightProducer.java:
```diff
@@ -181,5 +183,11 @@ void doAction(CallContext context, Action action,
     /** Whether the call has been cancelled by the client. */
     boolean isCancelled();
+
+    /** Get the middleware instance of the given type for this call. */
```
might be worth clarifying if this returns null or throws an exception if the middle was isn't found?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightProducer.java:
```diff
@@ -181,5 +183,11 @@ void doAction(CallContext context, Action action,
+    /** Get the middleware instance of the given type for this call. */
+    <T extends FlightServerMiddleware> T getMiddleware(FlightServerMiddleware.Key<T> key);
+
+    /** Get an immutable map of middleware for this call. */
```
is there a way to make the contract of immutability clear based on the return type?

lihalite replied:
Not without putting a Guava type in the public API, as Java collections don't have a separate type for immutable maps.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java:
```diff
@@ -247,6 +254,9 @@ public FlightServer build() {
         return null;
       });
 
+      if (!interceptors.isEmpty()) {
```
same question as above, is it really that much more efficient to avoid always creating?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java:
```diff
@@ -304,6 +314,20 @@ public Builder transportHint(final String key, Object option) {
     }
 
+    /**
+     * Add a Flight middleware to inspect and modify requests to this service.
```
```suggestion
     * Add a Flight middleware component to inspect and modify requests to this service.
```

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightServer.java:
```diff
@@ -304,6 +314,20 @@ public Builder transportHint(final String key, Object option) {
+     */
+    public <T extends FlightServerMiddleware> Builder middleware(final FlightServerMiddleware.Key<T> key,
+        final FlightServerMiddleware.Factory<T> factory) {
+      interceptors.add(new KeyFactory<T>(key, factory));
```
might pay to check that the key doesn't already exist?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightServerMiddleware.java:
```diff
@@ -0,0 +1,97 @@
+   * @param outgoingHeaders A mutable set of response headers. These can be manipulated to send different headers to the
+   *     client.
+   */
+  void sendingHeaders(CallHeaders outgoingHeaders);
```
onBeforeSendingHeaders...

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/ClientInterceptorAdapter.java:
```diff
@@ -0,0 +1,98 @@
+import io.grpc.Status;
+
+/**
+ * An adapter between Flight client middleware and gRPC interceptors.
```
probably worth documenting how this interacts with gRPC more explicitly.

lihalite replied:
I've filled out the docstring more.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/ContextPropagatingExecutorService.java:
```diff
@@ -0,0 +1,109 @@
+import io.grpc.Context;
+
+/**
+ * An {@link ExecutorService} that propagates the {@link Context}.
```
is Context thread safe?

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/ServerInterceptorAdapter.java:
```diff
@@ -0,0 +1,149 @@
+import io.grpc.Status;
+
+/**
+ * An adapter between Flight middleware and a gRPC interceptor.
```
please document the strategy it uses for interacting with gRPC?

lihalite replied:
Filled out the docstring more here.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/ServerInterceptorAdapter.java:
```diff
@@ -0,0 +1,149 @@
+
+    final SimpleForwardingServerCall<ReqT, RespT> forwardingServerCall = new SimpleForwardingServerCall<ReqT, RespT>(
+        call) {
+      // Not useful to provide since we don't want to expose the Protobuf. Instead, perform logging inside the method.
```
not sure how to interpret this comment?

emkornfield commented on java/flight/src/test/java/org/apache/arrow/flight/TestClientMiddleware.java:
```diff
@@ -0,0 +1,193 @@
+  }
+
+  @Test
+  public void injectSpans() {
```
this test could use some more documentation, it is a little sense to read as is.

lihalite replied:
I've renamed/cleaned up the tests here. The main scenario being tested is an OpenTracing-like example where a request ID gets invisibly propagated when you make a request.

emkornfield commented on java/flight/src/test/java/org/apache/arrow/flight/TestClientMiddleware.java:
```diff
@@ -0,0 +1,193 @@
+public class TestClientMiddleware {
+
+  @Test
+  public void rejectCall() {
```
tests are better named as either method_behaviortest_expectedOutcome or  behaviortest_expectedOutcome  if method isn't applicable.

lihalite commented:
Thanks @emkornfield for the thorough review. I'll address feedback and follow up.

lihalite commented:
Hopefully I've addressed feedback. I've also brought this up to par with C++/Python (only exposing text headers, using an enum to identify the RPC method being intercepted).

I will report back with benchmarks. As with C++, I don't expect much of a difference with middleware, but 1) the benchmark is broken and 2) the benchmarks we have measure the wrong thing (I would expect it to affect calls/s more than MB/s). 

Is there a path to getting Flight benchmarks into Archery, if they're just self-contained localhost ones?

lihalite commented:
Baseline on master (+patch to fix benchmarks):

```
Transferred 100000000 records totaling 3200000000 bytes at 722.680948 MiB/s. 23680809.290938 record/s. 5783.327245 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 957.121162 MiB/s. 31362946.231556 record/s. 7659.458729 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1052.723115 MiB/s. 34495631.022257 record/s. 8424.523008 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 622.971845 MiB/s. 20413541.408594 record/s. 4985.395083 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1095.041172 MiB/s. 35882309.111991 record/s. 8763.177531 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 998.438105 MiB/s. 32716819.837771 record/s. 7990.101741 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1030.980368 MiB/s. 33783164.684464 record/s. 8250.524479 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 994.156941 MiB/s. 32576534.631242 record/s. 7955.841288 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1014.400947 MiB/s. 33239890.240287 record/s. 8117.845994 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1005.636702 MiB/s. 32952703.465825 record/s. 8047.709240 batch/s.
```

Baseline on this branch:
```
Transferred 100000000 records totaling 3200000000 bytes at 585.023693 MiB/s. 19170056.375622 record/s. 4681.711168 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1000.070838 MiB/s. 32770321.223802 record/s. 8003.167849 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 973.966051 MiB/s. 31914919.552986 record/s. 7794.261653 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1001.385129 MiB/s. 32813387.920023 record/s. 8013.685598 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1026.597626 MiB/s. 33639551.004005 record/s. 8215.451146 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1015.019718 MiB/s. 33260166.123188 record/s. 8122.797771 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 634.883764 MiB/s. 20803871.185682 record/s. 5080.721421 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1020.964235 MiB/s. 33454956.060482 record/s. 8170.369369 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 624.611664 MiB/s. 20467275.011115 record/s. 4998.517903 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 614.653123 MiB/s. 20140953.546989 record/s. 4918.823675 batch/s.
```

If we just always enable interceptors:
```
Transferred 100000000 records totaling 3200000000 bytes at 585.145531 MiB/s. 19174048.762522 record/s. 4682.686189 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 615.882281 MiB/s. 20181230.578687 record/s. 4928.660132 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1066.380142 MiB/s. 34943144.490068 record/s. 8533.814747 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1095.946366 MiB/s. 35911970.517956 record/s. 8770.421440 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 640.732409 MiB/s. 20995519.571030 record/s. 5127.525790 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1039.859213 MiB/s. 34074106.707477 record/s. 8321.578340 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1085.222835 MiB/s. 35560581.843079 record/s. 8684.605298 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1081.660765 MiB/s. 35443859.949202 record/s. 8656.099477 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1060.792128 MiB/s. 34760036.434775 record/s. 8489.096098 batch/s.
Transferred 100000000 records totaling 3200000000 bytes at 1048.367279 MiB/s. 34352899.000164 record/s. 8389.664994 batch/s.
```

It seems max perf is about the same, but much less consistent. I'll remove the checks and always add the (dummy) interceptor.

emkornfield commented:
+1, thanks for th addressing the comments.  I reset the failing CI build, feel free to merge when everything is green.

lihalite commented:
Thanks @emkornfield!
